### ISSUE TITLE
vmadm's very slow "get" & "update" and option to ignore snapshots lookup

### ISSUE THREAD

**Issue by bassu** [joyent/smartos-live#282]
This might have already been observed but the `vmadm get` and updates become really slow when you have more than few snapshots for vm disks.

And it becomes somewhat of painful process when you are updating jsons for migration related purposes.

Clearly since this part of node and vmadm is slow, there should be an optional argument like `-i` or something to ignore looking up vm snapshots for `get` and `update`. 

```
# zfs list -t snapshot | wc -l
1093

# time zfs list -t snapshot | grep time
real    0m0.405s
user    0m0.113s
sys     0m0.296s

# time vmadm get 825e1f09-3339-4285-a480-89fe47af578e 1>/dev/null
real    0m14.738s
user    0m0.541s
sys     0m0.791s
```

**Comment by matthiasg** [comment]
@bassu why is this closed ? has this been addressed since ?

**Comment by bassu** [comment]
@matthiasg Yup, it was. I could no longer reproduce this issue starting early this year.
Probably they have started caching the <kbd>vmadm</kbd> output

For your *Quid pro quo*:

```
# zfs list -t snapshot | wc -l
    1724
# time vmadm list &>/dev/null
real    0m0.282s
user    0m0.182s
sys     0m0.099s
# time vmadm get 8a19636c-413a-4126-a2eb-67fa3a54c988 &>/dev/nul
real    0m0.398s
user    0m0.216s
sys     0m0.164s

**Comment by joshwilsdon** [comment]
From the output above, it looks like @bassu just has fewer snapshots now (1724 vs the previous ~6000) which has probably contributed to the performance improvement more than any changes in the software. I'm reopening this since I believe it is still actually a problem.

**Comment by joshwilsdon** [comment]
Looks like I might have misread the previous comments. The SmartOS system with 6000 was different than the system with the problem. That said, there haven't been changes to vmadm here and the problem still exists with large number of snapshots. So I think it's still correct to leave this open.

**Comment by bassu** [comment]
That's an interesting perspective indeed @joshwilsdon. 
My original post talks about 1000+ snapshots. Although I don't have the original machine with the same old set of configs/splatter/data on which I first encountered the problem; but I think the issue may be directly related to how fast ZFS lists the snapshots on varying amount of snapshot delta and segmentation of the pool.

IIRC, the actual library (I did patch it out first few times) behind `vmadm` has changed a lot since last 3+ years when I first reported the issue. Of course, the ZFS snapshotting code-base has also changed since then. Whatever the circumstance, I'm not able to reproduce the issue on new machines though. The longest I have had to wait for a `vmadm` listing during last couple of month was, probably 2 or 3 seconds.

**Comment by mgerdts** [comment]
For quite some time, ZFS lacked DSL caching, causing anything that interacted with the dataset layer (e.g. property lookups) to have to make a trip to disk.  This seems to be addressed with [OpenZFS 9337](https://github.com/zfsonlinux/zfs/commit/master), which landed in SmartOS with this changeset.

```
commit adb52d9262f45a04318fc6e188fe2b7f59d989a5
Author: Matthew Ahrens <mahrens@delphix.com>
Date:   Mon Jul 10 14:17:21 2017 -0700

    9337 zfs get all is slow due to uncached metadata
    Reviewed by: Prakash Surya <prakash.surya@delphix.com>
    Reviewed by: George Wilson <george.wilson@delphix.com>
    Reviewed by: Thomas Caputi <tcaputi@datto.com>
    Approved by: Richard Lowe <richlowe@richlowe.net>
```

Note that getting the value of a property on one dataset may require that an ancestor's dataset properties be looked up.  This is the case for most properties where a value has not been explicitly set on the dataset.
